tools/windows-node-installer/test/e2e/aws_test.go

```diff
@@ -1,90 +1,329 @@
-			return true
+			return
+		}
+	}
+	assert.Fail(t, "infrastructure tag not found")
+}
+
+// checkInstanceHasClusterWorkerSg asserts if the created instance has OpenShift cluster worker security group attached.
+func checkInstanceHasClusterWorkerSg(t *testing.T) {
+	workerSg, err := awsProvider.GetClusterWorkerSGID(infraID)
+	assert.NoError(t, err, "failed to get OpenShift cluster worker security group")
+
+	for _, sg := range createdInstance.SecurityGroups {
+		if *sg.GroupId == workerSg {
+			return
 		}
 	}
-	return false
+	assert.Fail(t, "instance is not associated with OpenShift cluster worker security group")
+}
+
+// checkInstanceIsAssociatedWithWindowsWorkerSg asserts if the created instance has a security group made for the
+// Windows instance attached by checking the group name, recorded id, necessary ports, and ip-permission.cidr.
+func checkInstanceIsAssociatedWithWindowsWorkerSg(t *testing.T) {
+	myIp, err := awscp.GetMyIp()
+	assert.NoError(t, err, "error getting user's public IP")
+
+	vpc, err := awsProvider.GetVPCByInfrastructure(infraID)
+	if err != nil {
+		assert.Fail(t, "error getting OpenShift cluster VPC, %v", err)
+		return
+	}
+
+	sgs, err := awsProvider.EC2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("ip-permission.from-port"),
+				Values: aws.StringSlice([]string{"-1", "3389"}),
+			},
+			{
+				Name:   aws.String("ip-permission.to-port"),
+				Values: aws.StringSlice([]string{"-1", "3389"}),
+			},
+			{
+				Name:   aws.String("ip-permission.protocol"),
+				Values: aws.StringSlice([]string{"tcp"}),
+			},
+			{
+				Name:   aws.String("group-id"),
+				Values: aws.StringSlice([]string{createdSgID}),
+			},
+			{
+				Name:   aws.String("ip-permission.cidr"),
+				Values: aws.StringSlice([]string{myIp + "/32", *vpc.CidrBlock}),
+			},
+		},
+	})
+	if err != nil || len(sgs.SecurityGroups) < 1 {
+		assert.Fail(t, "instance is not associated with a Windows security group, %v", err)
+	}
+}
+
+// checkInstanceIsAssociatedWithClusterWorkerIAM asserts if the created instance has the OpenShift cluster worker's IAM
+// attached.
+func checkInstanceIsAssociatedWithClusterWorkerIAM(t *testing.T) {
+	iamProfile, err := awsProvider.GetIAMWorkerRole(infraID)
+	assert.NoError(t, err, "error getting OpenShift Cluster Worker IAM")
+
+	assert.Equal(t, *iamProfile.Arn, *createdInstance.IamInstanceProfile.Arn, "instance is not associated with worker IAM profile")
+}
+
+// checkSgIsDeleted asserts if a security group is deleted by checking whether the security group exist on AWS.
+// If delete is successful, the id in createdSgID is erased.
+func checkSgIsDeleted(t *testing.T) {
+	sgs, err := awsProvider.EC2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		GroupIds: aws.StringSlice([]string{createdSgID}),
+	})
+
+	if err == nil && len(sgs.SecurityGroups) > 0 {
+		assert.Fail(t, "security group is not deleted")
+	} else {
+		createdSgID = ""
```

Comment:
Not clean but I can leave with it :)

Reply:
This is for deleting the id from the global variable if the sg is successfully deleted. I do not like how this is done either, but I wonder what could be the solution without an overkill.